mmdet/models/layers/conditional_transformer.py

```diff
@@ -0,0 +1,346 @@
+                self_attn_masks=None,
+                cross_attn_masks=None,
+                query_key_padding_mask=None,
+                key_padding_mask=None,
+                query_sine_embed=None,
+                is_first=None,
+                **kwargs):
+        query = self.self_attn(
+            query=query,
+            key=query,
+            query_pos=query_pos,
+            key_pos=query_pos,
+            attn_mask=self_attn_masks,
+            key_padding_mask=query_key_padding_mask,
+            **kwargs)
+        query = self.norms[0](query)
+        query = self.cross_attn(
+            query=query,
+            key=key,
+            query_pos=query_pos,
+            key_pos=key_pos,
+            attn_mask=cross_attn_masks,
+            key_padding_mask=key_padding_mask,
+            query_sine_embed=query_sine_embed,
+            is_first=is_first,
+            **kwargs)
+        query = self.norms[1](query)
+        query = self.ffn(query)
+        query = self.norms[2](query)
+
+        return query
+
+class ConditionalAttention(BaseModule):
+    """A wrapper of conditional attention, dropout and residual connection."""
+
+    def __init__(self,
+                 embed_dims,
+                 num_heads,
+                 dropout=0.,
+                 proj_drop=0.,
+                 batch_first=False,
+                 cross_attn=False,
+                 keep_query_pos=False,
+                 init_cfg=None):
+        super().__init__(init_cfg)
+        self.batch_first = batch_first  # indispensable
+        self.cross_attn = cross_attn
+        self.keep_query_pos = keep_query_pos
+        self.embed_dims = embed_dims  # output dims
+        self.num_heads = num_heads
+        self.attn_dropout = Dropout(dropout)
+        self.proj_drop = Dropout(proj_drop)
+
+        self._init_proj()
+
+    def _init_proj(self):
+        embed_dims = self.embed_dims
+        self.qcontent_proj = Linear(embed_dims, embed_dims)
+        self.qpos_proj = Linear(embed_dims, embed_dims)
+        self.kcontent_proj = Linear(embed_dims, embed_dims)
+        self.kpos_proj = Linear(embed_dims, embed_dims)
+        self.v_proj = Linear(embed_dims, embed_dims)
+        if self.cross_attn:
+            self.qpos_sine_proj = Linear(embed_dims, embed_dims)
+        self.out_proj = Linear(embed_dims, embed_dims)
+
+        nn.init.constant_(self.out_proj.bias, 0.)  # init out_proj
+
+    def forward_attn(self, query, key, value, attn_mask, key_padding_mask,
+                     need_weights):
+        assert key.size(0) == value.size(0), \
+            f'{"key, value must have the same sequence length"}'
+        assert query.size(1) == key.size(1) == value.size(1), \
+            f'{"batch size must be equal for query, key, value"}'
+        assert query.size(2) == key.size(2), \
+            f'{"q_dims, k_dims must be equal"}'
+        assert value.size(2) == self.embed_dims, \
+            f'{"v_dims must be equal to embed_dims"}'
```

Comment:
It looks vaguely familiar.
but why the prompt information in the assert message is deleted?
```Python
assert query.size(2) == key.size(2), \
    f'The q_dim {query.size(2)} must be equal to the k_dim' \
    f' {key.size(2)}'
assert value.size(2) == self.out_dims, \
    f'The v_dim {value.size(2)} must be equal to the out_dims ' \
    f'{self.out_dims}'
assert key.size(0) == value.size(0), \
    f'The k_len {key.size(0)} must be equal to the v_len ' \
    f'{value.size(0)}'
assert query.size(1) == key.size(1) == value.size(1), \
    f'The batch size (2nd) of q, k, v must be equal, ' \
    f'but got {query.size(1)}(q), {key.size(1)}(k), ' \
    f'{value.size(1)}(v)'
```